watchtower-plugin/src/retrier.rs

```diff
@@ -29,53 +31,61 @@ impl Retrier {
             wt_client,
             max_elapsed_time_secs,
             max_interval_time_secs,
+            pending_appointments: Arc::new(Mutex::new(HashMap::new())),
         }
     }
-    pub async fn manage_retry(&self, mut unreachable_towers: UnboundedReceiver<TowerId>) {
+    pub async fn manage_retry(
+        &self,
+        mut unreachable_towers: UnboundedReceiver<(TowerId, Locator)>,
+    ) {
         log::info!("Starting retry manager");
 
         loop {
-            let tower_id = unreachable_towers.recv().await.unwrap();
+            let (tower_id, locator) = unreachable_towers.recv().await.unwrap();
+            // Not start a retry if the tower is flagged to be abandoned
             {
-                // Not start a retry if the tower has been abandoned
-                let mut wt_client = self.wt_client.lock().unwrap();
+                let wt_client = self.wt_client.lock().unwrap();
+                if !wt_client.towers.contains_key(&tower_id) {
+                    continue;
+                }
                 if wt_client.towers.get(&tower_id).is_none() {
                     log::info!("Skipping retrying abandoned tower {}", tower_id);
                     continue;
                 }
-                wt_client.set_tower_status(tower_id, crate::TowerStatus::TemporaryUnreachable);
             }
 
+            self.add_pending_appointment(tower_id, locator);
+
             log::info!("Retrying tower {}", tower_id);
-            match retry_notify(
+            let r = retry_notify(
                 ExponentialBackoff {
                     max_elapsed_time: Some(Duration::from_secs(self.max_elapsed_time_secs as u64)),
                     max_interval: Duration::from_secs(self.max_interval_time_secs as u64),
                     ..ExponentialBackoff::default()
                 },
-                || async { self.add_appointment(tower_id).await },
+                || async { self.retry_tower(tower_id).await },
                 |err, _| {
                     log::warn!("Retry error happened with {}. {}", tower_id, err);
                 },
             )
-            .await
-            {
+            .await;
+
+            let mut state = self.wt_client.lock().unwrap();
+            self.pending_appointments.lock().unwrap().remove(&tower_id);
```

Comment:
I think we have a race condition here. Its scenario would be:
1- Tower A was temp unreachable and sent to be retried for more than once (say for 3 appointments)
2- It comes the turn for tower A to be retried for the first appointment and it fails and marks as unreachable (the problem here is that the appointment we just retried was never sent and also is removed from the pending appointments)
3- Tower A gets retried again (for the second appointment) and it succeeds (maybe it/we had a temporary internet issue), marking it as reachable.

In this scenario, the first appointment will never get sent to the tower (unless the user manually retry it).

Reply:
Umm, I don't follow. Here the whole tower record is removed from `pending_appointments`. This only happens if the whole retry strategy fails, so the data in the retrier is wipped.